Rewrite and reorganization of "Resolve" (#3)
LegReq/did-btcr2: chapters/CRUD-Operations.md

you must use to process a Beacon Signal for  a specific ::Beacon Type::.

| ::Beacon Type::      | Processing Algorithm                      |
|:---------------------|:------------------------------------------|
| ::Singleton Beacon:: | [Algo 9. Process Singleton Beacon Signal] |
| ::Map Beacon::       | [Algo 10. Process Map Beacon Signal]      |
| ::SMT Beacon::       | [Algo 11. Process SMT Beacon Signal]      |

Processing ::Beacon Signals:: may require ::Sidecar Data:: passed as part of the
resolution options in the DID resolution request.

The result of processing the ::Beacon Signals:: is a set of ::BTCR2 Updates::.
These updates must then be applied to the Contemporary DID Document using
[Algo 13. Apply BTCR2 Update].

#### Algo 9. Process Singleton Beacon Signal {.tabbed .unnumbered}

This algorithm processes a ::Beacon Signal:: broadcast from a ::Singleton
Beacon:: to retrieve and validate the ::BTCR2 update:: announced by this signal.

For signals from Singleton Beacons, the ::Signal Bytes:: are the SHA256 of a
canonicalized ::BTCR2 Update::. The BTCR2 update committed to by the Signal
Bytes must be retrieved, canonicalized using JCS, and then hashed. The resulting

> hash bytes must equal the Signal Bytes.

Reviewer: ```suggestion
hash bytes must equal the ::Signal Bytes::.
```

A BTCR2 Update may be retrieved from ::Sidecar Data::, or through querying a
::CAS::. If querying a ::CAS::, the ::Signal Bytes:: must be transformed into a
::CID:: according to the IPFS CID v1 specification. Inability to retrieve the
BTCR2 Update must raise a MISSING_UPDATE_DATA error.

##### Hide {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

Given the 32 `signalBytes` from the Singleton Beacon Signal and a
`sidecarDocumentsMap`:

1. Set `id` to the hexadecimal string representation of `signalBytes`.
2. Get `btcr2Update` from `sidecarDocumentsMap` by its `id` if available, or
   from ::CAS:: by its `id` if not and `cas` is defined.
3. If `btcr2Update` is undefined, raise InvalidDidUpdate error.
4. Set `btcr2Update`
5. Return `btcr2Update`.

NOTE. The act of retrieving from `sidecarDocumentsMap` or ::CAS:: validates the
document hash.

#### Algo 10. Process Map Beacon Signal {.tabbed .unnumbered}

This algorithm processes a ::Beacon Signal:: broadcast from a ::Map Beacon:: to
retrieve and validate the ::Beacon Announcement Map:: committed to by the
signal. That map is then used to retrieve and validate a ::BTCR2 Update:: using

> the specific identifier being resolved as the key.

Reviewer 2: ```suggestion
the SHA256 hash of the specific identifier being resolved as the key.
```

The ::Signal Bytes:: must be retrieved from the last transaction output of the
Beacon Signal. These bytes must then be used to retrieve and validate the Beacon
Announcement Map. A ::BTCR2 Update Announcement:: must be retrieved from the map
document using the hex encoded SHA256 hash of the identifier being resolved as

> the key. There may be no announcement for the identifier. The announcement is a

Reviewer 2: ```suggestion
the key. There MAY be no announcement for the identifier. The announcement is a
```

32-byte hash of the ::BTCR2 Update::, this must be used to retrieve and validate
the ::BTCR2 Update::.

Retrieval of the Beacon Announcement Map and ::BTCR2 Update:: documents may be
done through querying either ::Sidecar Data:: or a ::CAS::. If querying a
::CAS::, the 32-byte hashes must be transformed into a CID following the IPFS
CID v1 specification.

##### Hide {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

Given the 32 `signalBytes` from the Map Beacon Signal and a `sidecarDocumentsMap`:

1. Set `id` to the hexadecimal string representation of `signalBytes`.
2. Get `map` from `sidecarDocumentsMap` by its `id` if available, or from
3. ::CAS:: by its `id` if not and `cas` is defined.
3. If `map` is undefined, raise InvalidDidUpdate error.

> 4. Set `updateId` to the value of `map.<did>`.

Reviewer 2: ```suggestion
4. Set `updateId` to the value of `map.<did>`.
```

We need to add a step here that first hashes the did to generate the DID index in the map.
Probably easier to do in a IDE

Author: Done.

5. If `updateId` is undefined, return null.
6. Get `btcr2Update` from `sidecarDocumentsMap` by its `updateId` if available,
7. or from ::CAS:: by its `updateId` if not and `cas` is defined.

> 7. If `btcr2Update` is undefined, raise InvalidDidUpdate error.

Reviewer 2: ```suggestion
7. If `btcr2Update` is undefined, raise INVALID_DID_UPDATE error.
```

Author: Fixed in another commit.

       InvalidDidUpdate error.
9. If `verifyHashBytes` ≠ `signalBytes`, raise InvalidDidUpdate error.
10. If `updateId` is undefined, return null.
11. Get `btcr2Update` from `sidecarDocumentsMap` by its `updateId` if available,
    or from ::CAS:: by its `updateId` if not and `cas` is defined.
12. If `btcr2Update` is undefined, raise InvalidDidUpdate error.
13. Return `btcr2Update`.

NOTE. The act of retrieving from `sidecarDocumentsMap` validates the document
hash.

#### Algo 12. Derive Root Capability {.tabbed .unnumbered}

This algorithm derives a root capability to update a specific **did:btcr2**
identifier from the identifier itself. The root capability must be conformant to
the [Authorization Capabilities v0.3](https://w3c-ccg.github.io/zcap-spec/)
specification. The root capability for a specific **did:btcr2** identifier is a
JSON containing the following fields:

* @context: The string "[https://w3id.org/zcap/v1](https://w3id.org/zcap/v1)"
* id: A URN of the following format: urn:zcap:root:${encodeURIComponent(DID)}
* invocationTarget: The DID
* btcr2 identifiercontroller: The DID

Note: The controller of this capability is the DID, which resolves to a DID

> document. When verifying a BTCR2 update is an invocation of this capability, the

Reviewer 2: ```suggestion
document. When verifying a BTCR2 Update is an invocation of this capability, the
```

> controller resolves to the contemporary DID document. That is the DID document

Reviewer 2: ```suggestion
controller resolves to the Contemporary DID document. That is the DID document
```

that was current when the capability was invoked.

##### Hide {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

The algorithm takes in a did:btcr2 `identifier` and returns a `rootCapability`
object.

1. Define `rootCapability` as an empty object.
2. Set `rootCapability.@context` to ‘https://w3id.org/zcap/v1’.
3. Set `encodedIdentifier` to result of calling algorithm
   `encodeURIComponent(identifier)`.
4. Set `rootCapability.id` to `urn:zcap:root:${encodedIdentifier}`.
5. Set `rootCapability.controller` to `identifier`.
6. Set `rootCapability.invocationTarget` to `identifier`.
7. Return `rootCapability`.

#### Algo 13. Apply BTCR2 Update {.tabbed .unnumbered}

This algorithm takes a ::BTCR2 Update::, attempts to verify it and then applies
it to the appropriate DID document as identified by the update.

A ::BTCR2 Update:: must contain the following:

> * @context. A context array containing the follow context URLs

Reviewer 2: ```suggestion
* @context: A context array containing the follow context URLs
```

Feel like some find and replace caught this. We should review and fix. All these should be colon : not .

Author: Fixed, but it's still inconsistent with other parts of the document that use hyphens when documenting input parameters.

    * `"https://w3id.org/zcap/v1"`
    * `"https://w3id.org/security/data-integrity/v2"`
    * `"https://w3id.org/json-ld-patch/v1"`
    * `"https://btcr2.dev/context/v1"`

> * proof. A Data Integrity proof for with the proof purpose set to

Reviewer 2: ```suggestion
* proof. A Data Integrity proof produced using the [bip340-jcs-2025 cryptosuite](https://dcdpr.github.io/data-integrity-schnorr-secp256k1/) with the proof purpose set to
```

  capabilityInvocation. This must be an invocation of the capability to update
  the DID document of the DID being resolved. The root capability to update a
  specific **did:btcr2** identifier's DID document is derived following
  [Algo 12. Derive Root Capability] from **did:btcr2** identifier. While the
  [zCap-LD specification](https://w3c-ccg.github.io/zcap-spec) allows
  delegation, this specification does not define how such delegation might work.

> * patch. A [JSON patch](https://datatracker.ietf.org/doc/html/rfc6902) object

Reviewer 2: ```suggestion
* A [JSON patch](https://datatracker.ietf.org/doc/html/rfc6902) object
```

> * patch. A [JSON patch](https://datatracker.ietf.org/doc/html/rfc6902) object

Reviewer 2: ```suggestion
* patch: A [JSON patch](https://datatracker.ietf.org/doc/html/rfc6902) object
```

  that defines a set of transformations to be applied to a DID document. The
  result of applying the patch must be a conformant DID document according to
  the DID core v1.1 specification.
* targetVersionId. The versionId of the DID document after the patch has been
  applied. The patch must be applied to the DID document with a versionId that

>   is one less than the targetVersionId. If the contemporary document is equal to

Reviewer 2: ```suggestion
  is one less than the targetVersionId. If the ::Contemporary DID Document:: is equal to
```

  or less than the targetVersionId, then the ::Unsecured BTCR2 Update:: must be
  the same as the previously applied ::BTCR2 Update::. If the targetVersionId is

>   greater than one plus the contemporary document’s versionId, then a

Reviewer 2: ```suggestion
  greater than one plus the Contemporary DID Document’s versionId, then a
```

(CAS) such as IPFS. If retrieving from ::CAS::, the ::Genesis Bytes:: must be
transformed into a ::content identifier:: according to the CID v1.0
specification.

If the ::Genesis Document:: cannot be retrieved, it results in a FILE_NOT_FOUND.
If the ::Initial DID Document:: is not conformant, then it results in an
INVALID_DID error.

#### Algo 7. Verify Genesis Document {.tabbed .unnumbered}

Verifying the ::Genesis Document:: requires checking its hash against the
::Genesis Bytes:: and then validating that it is a conformant document.

To check its hash, compute the SHA256 of the ::Genesis Document:: canonicalized
using the [JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785)
(JCS). The computed hash must be equal to the ::Genesis Bytes::. If it is not,
this is an INVALID_DID_DOCUMENT_CONSTRUCTION error.

To check that it is conformant, transform the ::Genesis Document:: into the
::Initial DID Document:: by replacing the placeholder DID with the **did:btcr2**
identifier being resolved. The transformed document must be a conformant DID
document according to the DID core v1.1 specification. If it is not, this is an

> INVALID_DID_DOCUMENT error

Reviewer 2: ```suggestion
INVALID_DID_DOCUMENT error.
```

> Algo 8. Find & Process Beacon Signals
> 
> 1. Algo 9. Scan until Signal, Select Block
> 2. Algo 10. For block, process Signals
>     1. For each signal,
>         1. Algo 11,12,13 Beacon Specific Algo
>         2. Algo 14 Apply Update
> 3. Continue Scanning until end state

Reviewer 2: ```suggestion
```

This needs to go

this document according to [Algo 7. Verify Genesis Document].

Second, iterate through ::Beacon Signals:: posted to the Bitcoin blockchain
using [Algo 8. Process Beacon Signals] processing them according to their
::Beacon Type::. Processing ::Beacon Signals:: retrieves the announced ::BTCR2
Update:: for the DID being resolved. The algorithms for processing ::Beacon
Signals:: according to their ::Beacon Types:: are as follows: for ::Singleton
Beacons:: use [Algo 9. Process Singleton Beacon Signal];  for ::Map Beacons::
use [Algo 10. Process Map Beacon Signal]; for ::SMT Beacons:: use [Algo 11.
Process SMT Beacon Signal].

::BTCR2 Updates:: announced with ::Beacon Signals:: are then applied to the DID
document. ::BTCR2 Updates:: MUST be invocations of the capability to update the
DID being resolved. To verify this, the resolver MUST first derive the root
capability from the DID itself using  [Algo 12. Derive Root Capability], then
verify and apply the invocation using [Algo 13. Apply BTCR2 Update].

Traverse the blockchain and apply updates in temporal order, updating and
maintaining a ::Contemporary DID Document:: until all updates are processed.

> At each stage, the Contemporary DID Document defines the ::BTCR2 Beacons::

Reviewer: ```suggestion
At each stage, the ::Contemporary DID Document:: defines the ::BTCR2 Beacons::
```

this table.

| Value | Network          |
|:------|:-----------------|
| 0     | bitcoin          |
| 1     | signet           |
| 2     | regtest          |
| 3     | testnet3         |
| 4     | testnet4         |
| 5     | mutinynet        |
| C     | custom network 1 |
| D     | custom network 2 |
| E     | custom network 3 |
| F     | custom network 4 |

Any value not in the table results in an INVALID_DID Error.

To validate the ::Genesis Bytes::, check the hrp value. If hrp is ‘k’, the
::Genesis Bytes:: must be a 33-byte representation of a compressed secp256k1
public key. If hrp is ‘x’, the ::Genesis Bytes:: must be a 32-byte SHA256 hash
of a ::Genesis Document::. If these conditions are not met, or hrp has any other
value, it results in an INVALID_DID Error.

Then the resolution options of a resolution request must be processed and the
following fields must be used to execute the request:

* versionId: The versionId of the DID document targeted by this resolution
  request. If there is a DID document with this versionId, it MUST be returned
  in the resolution result.
* versionTime: A UTC timestamp that is targeted by the resolution request. The

>   resolve MUST return the contemporary DID document at this time as determined

Reviewer: ```suggestion
  resolve MUST return the Contemporary DID Document at this time as determined
```

>   resolve MUST return the Contemporary DID Document at this time as determined

Reviewer: ```suggestion
  resolve MUST return the ::Contemporary DID Document:: at this time as determined
```

(JCS). The computed hash MUST be equal to the ::Genesis Bytes::. If it is not,
this is an INVALID_DID_DOCUMENT_CONSTRUCTION error.

To check that it is conformant, transform the ::Genesis Document:: into the
::Initial DID Document:: by replacing the placeholder DID with the **did:btcr2**
identifier being resolved. The transformed document MUST be a conformant DID
document according to the [DID core v1.1](https://www.w3.org/TR/did-1.1/) specification. If it is not, 
the algorithm MUST result in an INVALID_DID_DOCUMENT error.

#### Algo 8. Process Beacon Signals {.tabbed .unnumbered}

Traverse the Bitcoin blockchain from the start of the chain and identify the first
Bitcoin block that contains a ::Beacon Signal::. For all ::Beacon Signals:: in
that block, process these signals to retrieve ::BTCR2 Updates:: announced by
these signals. Apply these updates to the DID document to produce the
::Contemporary DID Document:: with that block. Continue traversing the
blockchain, processing blocks with signals, until the latest block.

A ::Beacon Signal:: is a Bitcoin transaction that spends at least one ::UTXO::
controlled by a ::Beacon Address:: defined by a Beacon in the service array of

> the Contemporary DID Document. The Beacon Signal contains ::Signal Bytes::, 32

Reviewer: ```suggestion
the ::Contemporary DID Document::. The ::Beacon Signal:: contains ::Signal Bytes::, 32
```

bytes included in the last transaction output of the transaction. The script of
this transaction output MUST have the following form: `[OP_RETURN, OP_PUSH_BYTES,
<32 signal bytes>]`.

Process ::Beacon Signals:: according to their type, as identified by the Beacon

> Service in the Contemporary DID Document. The table below defines the algorithm

Reviewer: ```suggestion
Service in the ::Contemporary DID Document::. The table below defines the algorithm
```

10. If `updateId` is undefined, return null.  
11. Get `btcr2Update` from `sidecarDocumentsMap` by its `updateId` if available, 
or from [CAS](https://dcdpr.github.io/did-btc1/#def-content-addressable-storage) by 
its `updateId` if not and `cas` is defined.  
12. If `btcr2Update` is undefined, raise a MISSING\_UPDATE\_DATA error.  
13. Return `btcr2Update`.

#### Algo 12. Derive Root Capability {.tabbed .unnumbered}

This algorithm derives a root capability to update a specific **did:btcr2**
identifier from the identifier itself. The root capability must be conformant to
the [Authorization Capabilities v0.3](https://w3c-ccg.github.io/zcap-spec/)
specification. The root capability for a specific **did:btcr2** identifier is a
JSON containing the following fields:

* \@context: The string "[https://w3id.org/zcap/v1](https://w3id.org/zcap/v1)"
* id: A URN of the following format: urn:zcap:root:${encodeURIComponent(DID)}
* invocationTarget: The DID
* btcr2 identifiercontroller: The DID

Note: The controller of this capability is the DID, which resolves to a DID
document. When verifying a BTCR2 Update is an invocation of this capability, the

> controller resolves to the Contemporary DID document. That is the DID document

Reviewer: ```suggestion
controller resolves to the ::Contemporary DID Document::. That is the DID document
```

* proof: A Data Integrity proof for with the proof purpose set to
  capabilityInvocation. This must be an invocation of the capability to update
  the DID document of the DID being resolved. The root capability to update a
  specific **did:btcr2** identifier's DID document is derived following
  [Algo 12. Derive Root Capability] from **did:btcr2** identifier. While the
  [zCap-LD specification](https://w3c-ccg.github.io/zcap-spec) allows
  delegation, this specification does not define how such delegation might work.
* patch: A [JSON patch](https://datatracker.ietf.org/doc/html/rfc6902) object
  that defines a set of transformations to be applied to a DID document. The
  result of applying the patch must be a conformant DID document according to
  the DID core v1.1 specification.
* targetVersionId: The versionId of the DID document after the patch has been
  applied. The patch must be applied to the DID document with a versionId that
  is one less than the targetVersionId. If the ::Contemporary DID Document:: is equal to
  or less than the targetVersionId, then the ::Unsecured BTCR2 Update:: must be
  the same as the previously applied ::BTCR2 Update::. If the targetVersionId is

>   greater than one plus the Contemporary DID Document’s versionId, then a

Reviewer: ```suggestion
  greater than one plus the ::Contemporary DID Document::’s versionId, then a
```

I checked; the :: followed by an apostrophe does parse correctly

##### Example code {.unnumbered .unlisted}

### Resolve

Resolving a **did:btcr2** identifier takes an input DID and returns the
canonical DID document at the requested target time as well as additional

> metadata in a DID Resolution Result. The current time is the default.

Reviewer 2: ```suggestion
metadata in a DID Resolution Result. The current time is the default target time for a resolution request.

That sentence just feels hanging and odd otherwise to me
```

Author: Fixed in a separate commit, as your comment is included in the suggested change section.

Resolution takes part in two steps.

First, establish the ::Initial DID Document::. Start processing the resolution
inputs, parsing the identifier and any resolution options using [Algo 4. Process
Resolution Inputs]. Then establish the ::Initial DID Document:: according to the
type of the identifier. If the identifier encodes a 33-byte compressed secp256k1
public key, the ::Initial DID Document:: can be generated by following [Algo 5.
Deterministically Generate Initial DID Document]. Otherwise, the identifier must
encode a 32-byte hash of a ::Genesis Document::, in this case first retrieve the

> ::Genesis Document:: using [Algo 6. Retrieve Genesis Document] and then verify

Reviewer: ```suggestion
::Genesis Document:: using [Algo 6. Retrieve Genesis DID Document] and then verify
```

Correct Algo name

> this document according to [Algo 7. Verify Genesis Document].

Reviewer 2: Hmm I think I would go the over way. Genesis Document is defined in the terms. Lets change the algo names in the header

Second, iterate through ::Beacon Signals:: posted to the Bitcoin blockchain
using [Algo 8. Process Beacon Signals] processing them according to their

> ::Beacon Type::. Processing Beacon Signals retrieves the announced ::BTCR2

Reviewer: If Processing Beacon Signals is referring to Algo 8, 
```suggestion
::Beacon Type::. Process Beacon Signals retrieves the announced ::BTCR2
```
Otherwise Beacon Signals should be a reference

Author: I don't understand this change. The sentence is fine as is, in my opinion.

Reviewer: If we are referring specifically to Algo 8, the name should match. If not, 
```suggestion
::Beacon Type::. Processing ::Beacon Signals:: retrieves the announced ::BTCR2
```

That said, I don't feel strongly about it either way or even being done at all.

Reviewer 2: We are not referring to the algo here. Adding the :: makes sense though

Update:: for the DID being resolved. The algorithms for processing ::Beacon
Signals:: according to their ::Beacon Types:: are as follows: for ::Singleton
Beacons:: use [Algo 9. Process Singleton Beacon Signal];  for ::Map Beacons::
use [Algo 10. Process Map Beacon Signal]; for ::SMT Beacons:: use [Algo 11.
Process SMT Beacon Signal].

::BTCR2 Updates:: announced with ::Beacon Signals:: are then applied to the DID
document. ::BTCR2 Updates:: must be invocations of the capability to update the
DID being resolved. To verify this, the resolver must first derive the root
capability from the DID itself using  [Algo 12. Derive Root Capability], then

> verify and apply the invocation using [Algo 13. Apply BTCR Update].

Reviewer: ```suggestion
verify and apply the invocation using [Algo 13. Apply BTCR2 Update].
```

Correct Algo name

Traverse the blockchain and apply updates in temporal order, updating and

> maintaining a ::Contemporary DID Document:: until all updates are processed.

Reviewer: ```suggestion
maintaining a Contemporary DID Document until all updates are processed.
```
Or add Contemporary DID Document to terminology and update it. Maybe create an issue for that?

Reviewer 2: We need to add Contemporary DID Document to the terminology. Lets do that in this PR I think

At each stage, the Contemporary DID Document defines the ::BTCR2 Beacons::

> active at that point in time; those beacons–and only those Beacon–are checked

Reviewer: ```suggestion
active at that point in time; those Beacons – and only those Beacons – are checked
```
Spaces around the dashes, fix case, match plurals

> active at that point in time; those beacons–and only those Beacon–are checked

Reviewer 2: ```suggestion
active at that point in time; those beacons–and only those beacon–are checked
```

think not capitalized?

for updates.

Continue traversing the chain until all timely, authentic ::Beacon Signals::
have been processed and their announced updates have been applied in order.

The resulting DID document is the canonical DID Document and must be returned to
the caller of the resolver function in a resolution response which can be
constructed following [Algo 14. Construct Resolution Result].

> NOTE. If there are no updates, the ::Initial DID Document:: is returned.

Reviewer 2: ```suggestion
NOTE. If there are no updates, the ::Initial DID Document:: is returned unless a specific versionId has been targetted by the resolution request.
```

#### Algo 4. Process Resolution Inputs {.tabbed .unnumbered}

This algorithm processes the resolution inputs from a resolution request. First,
parse the identifier string, then process the resolution options.

First take the identifier string, validate it according to the DID syntax and
decode it to retrieve the identifier type, version, network and ::Genesis
Bytes::.

The identifier string must consist of three components separated by a ‘:’
character. The first component must be the string ‘did’, the second component
must be ‘btcr2’ and the third component is the BTCR2-specific identifier.
Otherwise an INVALID_DID error MUST be raised.

> The BTCR2-specific identifier is decoded using the bech32m algorithm which

Reviewer: ```suggestion
The BTCR2-specific identifier is decoded using the [bech32m](https://en.bitcoin.it/wiki/BIP_0350#Bech32m) algorithm which
```

Add link for bech32m

> The BTCR2-specific identifier is decoded using the bech32m algorithm which

Reviewer 2: ```suggestion
The BTCR2-specific identifier is decoded using the bech32m algorithm which
```

Should we cite the bech32m algo again here?

| 1     | signet           |
| 2     | regtest          |
| 3     | testnet3         |
| 4     | testnet4         |
| 5     | mutinynet        |
| C     | custom network 1 |
| D     | custom network 2 |
| E     | custom network 3 |
| F     | custom network 4 |

Any value not in the table results in an INVALID_DID Error.

To validate the ::Genesis Bytes::, check the hrp value. If hrp is ‘k’, the
::Genesis Bytes:: must be a 33-byte representation of a compressed secp256k1
public key. If hrp is ‘x’, the ::Genesis Bytes:: must be a 32-byte SHA256 hash
of a ::Genesis Document::. If these conditions are not met, or hrp has any other
value, it results in an INVALID_DID Error.

Then the resolution options of a resolution request must be processed and the
following fields must be used to execute the request:

* versionId: The versionId of the DID document targeted by this resolution
  request. If there is a DID document with this versionId, it MUST be returned
  in the resolution result.
* versionTime: A UTC timestamp that is targeted by the resolution request. The
  resolve MUST return the contemporary DID document at this time as determined
  by the Bitcoin block time.

> * verbose: A boolean value. This client is requesting additional detail as to 

Reviewer: Is it up to the implementation to decide what that additional detail is?

Author: Yes, I would think so.

Reviewer 2: We talk to it a little bit in the construct Resolution Response. I wasnt sure how much detail to go into

  how the resolution result was constructed.
* sidecar: The set of ::Sidecar Data:: provided to the Resolver by the client as
  part of their resolution request. This data is likely provided to the client
  by the DID controller.
  * documents: An array of JSON documents that must be used to execute the
    resolution request. ::Sidecar:: documents are identified within ::Beacon
    Signals:: by their hash, this should be used to retrieve the relevant
    document from ::Sidecar Data::.
  * smtProofs: An array of SMT proofs that prove inclusion or non-inclusion of a

>     ::BTCR2 Update:: announced within a SMT Beacon Signal. Each proof must have

Reviewer: ```suggestion
    ::BTCR2 Update:: announced within an SMT Beacon Signal. Each proof must have
```

Reviewer 2: Really, for me it is a SMT Beacon Signal

    an id field which is the hex encoded merkle root that is included as the

>     Signal Bytes of the Beacon Signal.

Reviewer 2: ```suggestion
    Signal Bytes of a SMT Beacon Signal.
```

##### Hide {.unnumbered .unlisted}

##### Flowchart {.unnumbered .unlisted}

##### Examples {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

Given:

* `identifier` - required, a string **did:btcr2** identifier

Decode the **did:btcr2** identifier as follows:

1. Split `identifier` into an array of `components` at the colon `.` character.
2. If the length of the `components` array is not `3`, raise `invalidDid` error.
3. If `components[0]` is not "did", raise `invalidDid` error.
4. If `components[1]` is not "btcr2", raise `methodNotSupported` error.
5. Set `encodedString` to `components[2]`.

> 6. Pass `encodedString` to the [Bech32m Decoding] algorithm, retrieving `hrp`

Reviewer 2: ```suggestion
6. Decode the `encodedString` using the Bech32m Decoding algorithm defined in [BIP350](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki), retrieving `hrp`
```

Reviewer 2: My sense is we are going to removed the Bech32m algorithm. I missed this. 

Note to all: as we are reviewing we should be especially careful of any calls to other algorithms. These likely are incorrect

   and `dataBytes`.

> 7. If the [Bech32m Decoding] algorithm fails, raise `invalidDid` error.

Reviewer 2: ```suggestion
7. If the decoding algorithm fails, raise `INVALID_DID` error.
```

8. Map `hrp` to `idType` from the following:
   1. "k" - "key"
   2. "x" - "external"

>    3. other - raise `invalidDid` error

Reviewer 2: ```suggestion
   3. other - raise `INVALID_DID` error
```

Seems I missed a lot of error updates.

9. Set `version` to `1`.
10. If at any point in the remaining steps there are not enough nibbles to

>     complete the process, raise `invalidDid` error.

Reviewer 2: ```suggestion
    complete the process, raise `INVALID_DID` error.
```

11. Start with the first nibble (the higher nibble of the first byte) of
    `dataBytes`.
12. Add the value of the current nibble to `version`.
13. If the value of the nibble is hexadecimal `F` (decimal `15`), advance to the
    next nibble (the lower nibble of the current byte or the higher nibble of
    the next byte) and return to the previous step.

> 14. If `version` is greater than `1`, raise `invalidDid` error.

Reviewer 2: ```suggestion
14. If `version` is greater than `1`, raise `INVALID_DID` error.
```

15. Advance to the next nibble and set `networkValue` to its value.
16. Map `networkValue` to `network` from the following:
    1. `0` - "bitcoin"
    2. `1` - "signet"
    3. `2` - "regtest"
    4. `3` - "testnet3"
    5. `4` - "testnet4"
    6. `5` - "mutinynet"

>     7. `6`-`B` - raise `invalidDid` error

Reviewer 2: ```suggestion
    7. `6`-`B` - raise `INVALID_DID` error
```

    8. `C`-`F` - `networkValue - 11`
17. If the number of nibbles consumed is odd:
    1. Advance to the next nibble and set `fillerNibble` to its value.
    2. If `fillerNibble` is not `0`, raise `invalidDid` error.
18. Set `genesisBytes` to the remaining `dataBytes`.
19. If `idType` is "key" and `genesisBytes` is not a valid compressed secp256k1

>     public key, raise `invalidDid` error.

Reviewer 2: ```suggestion
    public key, raise `INVALID_DID` error.
```

20. Return `idType`, `version`, `network`, and `genesisBytes`.

##### Example code {.unnumbered .unlisted}

##### Links {.unnumbered .unlisted}

to btcr2.dev discussion forum

#### Algo 5. Deterministically Generate Initial DID Document {.tabbed .unnumbered}

This algorithm deterministically generates an ::Initial DID Document:: from a
33-byte compressed representation of a secp256k1 public key encoded into the
identifier as the ::Genesis Bytes::. The DID document generated must be a
conformant DID document according to the DID core v1.1 specification.

The DID document must contain only the following properties:

> * A @context array containing the DID core v1.1 context url ("[https://www.w3.org/ns/did/v1.1](https://www.w3.org/ns/did/v1.1)" )

Reviewer: `@context` is not parsing correctly. 
This is line showing up as "A 1 array containing..."

Author: Requires a leading '\\'. Fixed in another commit.

  and the did:btcr2 context ("[https://btcr2.dev/context/v1](https://btc1.dev/context/v1)").
* A verificationMethod array containing a single verificationMethod. This
  verificationMethod must have an id value of "#initialKey" be of the type
  Multikey with the publicKeyMultibase value a multikey encoding of the
  secp256k1 public key bytes.

> * An array containing the verificationMethod id ‘#initialKey’ for each of the

Reviewer: ```suggestion
* An array containing the verificationMethod id "#initialKey" for each of the
```

Quote type consistency

  verification relationships. assertionMethod, authentication,
  capabilityInvocation, capabilityDelegation
* A service array containing three BTCR2 Beacon services. The serviceEndpoint
  for each of these services must be a BIP21 URI encoding of a bitcoin address

>   controlled by the public key. The three different address formats are. p2pkh,

Reviewer: ```suggestion
  controlled by the public key. The three different address formats are p2pkh,
```
delete extra period

  p2wpkh, p2tr. The id value of the service must be "#initialP2PKH",
  #initialP2WPKH” and "#initialP2TR" respectively. Finally, each of these

>   services must have a type set to the array \["BTCR2Beacon",
>   "SingletonBeacon"\]

Reviewer 2: ```suggestion
  services must have a type set to the string "SingletonBeacon"
```

We are not handling the additional beacon types in this PR. That is a discussion we want to have in person.

For now we just want to document what we have implemented in the code as much as possible

   `capabilityDelegation` properties in `initialDocument` to a copy of the
   `tempArray` variable.
10. Set the `initialDocument.services` property in `initialDocument` to the
    result of passing the `identifier`, `keyBytes` and
    `identifierComponents.network` to the [Deterministically Generate Beacon
    Services] algorithm.
11. Return `initialDocument`.

##### Example code {.unnumbered .unlisted}

##### Links {.unnumbered .unlisted}

to btcr2.dev discussion forum

#### Algo 6. Retrieve Genesis DID Document {.tabbed .unnumbered}

This algorithm uses the ::Genesis Bytes:: encoded into an external **did:btcr2**
identifier to retrieve and verify the ::Genesis Document::. This document is
then transformed into the ::Initial DID Document::, which must be validated as a
conformant DID document according to the DID core v1.1 specification.

The ::Genesis Document:: may be retrieved either from the ::Sidecar Data::
provided in the DID resolution options or from a ::Content Addressable Storage::
(CAS) such as IPFS. If retrieving from ::CAS::, the ::Genesis Bytes:: must be
transformed into a ::content identifier:: according to the CID v1.0
specification.

> If the ::Genesis Document:: cannot be retrieved, it results in a FILE_NOT_FOUND.

Reviewer: ```suggestion
If the ::Genesis Document:: cannot be retrieved, it results in a FILE_NOT_FOUND error.
```

using the [JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785)
(JCS). The computed hash must be equal to the ::Genesis Bytes::. If it is not,
this is an INVALID_DID_DOCUMENT_CONSTRUCTION error.

To check that it is conformant, transform the ::Genesis Document:: into the
::Initial DID Document:: by replacing the placeholder DID with the **did:btcr2**
identifier being resolved. The transformed document must be a conformant DID
document according to the DID core v1.1 specification. If it is not, this is an
INVALID_DID_DOCUMENT error

##### Hide {.unnumbered .unlisted}

##### Flowchart {.unnumbered .unlisted}

##### Examples {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

##### Example code {.unnumbered .unlisted}

##### Links {.unnumbered .unlisted}

to btcr2.dev discussion forum

Algo 8. Find & Process Beacon Signals

1. Algo 9. Scan until Signal, Select Block
2. Algo 10. For block, process Signals
    1. For each signal,
        1. Algo 11,12,13 Beacon Specific Algo
        2. Algo 14 Apply Update
3. Continue Scanning until end state

#### Algo 8. Process Beacon Signals {.tabbed .unnumbered}

> Traverse the Bitcoin blockchain from the start of the chain, identify the next

Reviewer: ```suggestion
Traverse the Bitcoin blockchain from the start of the chain and identify the first
```
"first" seems to make more sense here.

Bitcoin block that contains a ::Beacon Signal::. For all ::Beacon Signals:: in
that block, process these signals to retrieve ::BTCR2 Updates:: announced by
these signals. Apply these updates to the DID document to produce the

> ::Contemporary DID Document:: with that block. Continue traversing the

Reviewer: ```suggestion
Contemporary DID Document with that block. Continue traversing the
```
See comment on line 244

> blockchain until the latest block.

Reviewer: ```suggestion
blockchain, processing blocks with signals, until the latest block.
```

provided in the DID resolution options or from a ::Content Addressable Storage::
(CAS) such as IPFS. If retrieving from ::CAS::, the ::Genesis Bytes:: must be
transformed into a ::content identifier:: according to the CID v1.0
specification.

If the ::Genesis Document:: cannot be retrieved, it results in a FILE_NOT_FOUND.
If the ::Initial DID Document:: is not conformant, then it results in an
INVALID_DID error.

##### Hide {.unnumbered .unlisted}

##### Flowchart {.unnumbered .unlisted}

##### Examples {.unnumbered .unlisted}

##### Imperative Algorithm {.unnumbered .unlisted}

##### Example code {.unnumbered .unlisted}

##### Links {.unnumbered .unlisted}

to btcr2.dev discussion forum

#### Algo 7. Verify Genesis Document {.tabbed .unnumbered}

Verifying the ::Genesis Document:: requires checking its hash against the
::Genesis Bytes:: and then validating that it is a conformant document.

To check its hash, compute the SHA256 of the ::Genesis Document:: canonicalized
using the [JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785)
(JCS). The computed hash must be equal to the ::Genesis Bytes::. If it is not,

> this is an INVALID_DID_DOCUMENT_CONSTRUCTION error.

Reviewer 2: ```suggestion
this is an INVALID_GENESIS_DOCUMENT error.
```

What about this?

Author: That's specific to there being an invalid genesis document. I was thinking that a missing or invalid genesis document or BTCR2 update should generate the same error, hence "INVALID_DID_DOCUMENT_CONSTRUCTION".

However, if we go with the "hash everything in the array and use the hash as the key to retrieve the document", all we will get is a "file not found" error, as we won't know which of the documents in the map is the genesis document without inspecting them individually.

Reviewer 2: I think this: "hash everything in the array and use the hash as the key to retrieve the document"

Is an implementation detail not something we would specify. Even though i agree that is the way you should do it.

Reviewer 2: This algorithm is assuming you have a Genesis Document "from somewhere". We need to verify it is the right one,

Reviewer 2: I think we need to not get bogged down in what the error is for now. Maybe we just leave as is and can revisit errors later.
